library/core/src/fmt/rt.rs

```diff
@@ -162,7 +159,14 @@ impl<'a> Argument<'a> {
     #[inline(always)]
     pub(super) unsafe fn fmt(&self, f: &mut Formatter<'_>) -> Result {
         match self.ty {
-            ArgumentType::Placeholder { formatter, value } => formatter(value, f),
+            // SAFETY:
+            // `Argument` is constructed so that if `formatter` originally had
+            // the type `fn(&T, ...)` then `value` has type `&T`. Since we use
+            // `value` within the lifetime 'a of the reference and references
+            // and `NonNull` are ABI-compatible, this is completely equivalent
+            // to calling the original function passed to `new` with the original
+            // reference, which is always sound.
```

Comment:
This is actually a data structure invariant, isn't it?
So ideally the reasoning would be: have that invariant documented in the type, and then in `Argument::new` justify why are are establishing the invariant.

Currently this makes an argument about code somewhere else, which is non-local -- that's not great.

Reply:
Fair point, should be fine now.